COLLECTIONS: add option for local, cloud, or openai collections

mclean-connor opened the pull request.

innov8r-44 commented on packages/server/src/utils/CollectionsHelper.ts:
```diff
@@ -0,0 +1,100 @@
+
+export const createFiles = async (collection: Collection, names: string[], dataSource: DataSource) => {
+    const results = []
+    for (const name of names) {
+        const newFile = new CollectionFile()
+        Object.assign(newFile, { fileName: name })
+        newFile.collection = collection as Collection
+
+        const file = dataSource.getRepository(CollectionFile).create(newFile)
+        const result = await dataSource.getRepository(CollectionFile).save(file)
+        results.push(result)
+    }
```
Can we collect the results and turn them to a Promise.all? This would only work if there aren't a ton of files being created at once, as that could cause issue trying to run them all at once. Here is a possible implementation using Promise.all:

```
export const createFiles = async (collection: Collection, names: string[], dataSource: DataSource) => {
    const fileCreationPromises = names.map(async (name) => {
        const newFile = new CollectionFile();
        Object.assign(newFile, { fileName: name });
        newFile.collection = collection;

        const file = dataSource.getRepository(CollectionFile).create(newFile);
        return dataSource.getRepository(CollectionFile).save(file);
    });

    const results = await Promise.all(fileCreationPromises);
    return results;
}
```

mclean-connor replied:
Yup this should work!

mclean-connor replied:
I'll need to update changes like this in a few other places also

innov8r-44 commented on packages/server/src/utils/openAiHelpers.ts:
```diff
@@ -0,0 +1,94 @@
+        const collectionFiles = []
+        for (const f of Array.from(files.data)) {
+            if (f.filename.includes(name) && f.filename.includes(String(user.id))) {
+                const name = f.filename.split('_').pop()
+                collectionFiles.push({ fileName: name, langchain_primaryid: f.id })
```
we may be able to combine the following, I personally like having it in two lines but I know some people prefer to not define single use variables, here is what I am thinking:

```
collectionFiles.push({ fileName: f.filename.split('_').pop(), langchain_primaryid: f.id })
```
this way we also don't reuse `name`. 

The other option here could be:

 const filename = f.filename.split('_').pop()
 collectionFiles.push({ fileName, langchain_primaryid: f.id })

innov8r-44 commented on packages/server/src/utils/openAiHelpers.ts:
```diff
@@ -0,0 +1,94 @@
+        return addedFiles
+    }
+
+    async deleteFile(user: User, fileIds: string[]) {
+        const deleted: FileDeleted[] = []
+        for (const id of fileIds) {
+            const del = await this.client.files.del(id)
+            deleted.push(del)
+        }
+
+        return deleted
+    }
```
Possible use of `Promose.all()`:

```
async deleteFile(user: User, fileIds: string[]) {
    const deletePromises = fileIds.map(id => this.client.files.del(id));

    try {
        const deleted = await Promise.all(deletePromises);
        return deleted;
    } catch (e) {
        console.error(e);
        // Handle or rethrow the error as appropriate
        throw e;
    }
}

```

innov8r-44 commented on packages/server/src/utils/openAiHelpers.ts:
```diff
@@ -0,0 +1,94 @@
+        return { data: collectionFiles }
+    }
+
+    async addFiles(user: User, collection: string, fileData: any) {
+        // the collection name and user id are added to the file name
+
+        const addedFiles = []
+        for (const f of fileData) {
+            const splitDataURI = f.split(',')
+            const fileName = getFileName(f)
+            const updatedName = `${collection}_${String(user.id)}_${fileName}`
+
+            splitDataURI.pop()
+            const bf = Buffer.from(splitDataURI.pop() || '', 'base64')
+            const cf = new File([bf], updatedName)
+            let file
+            try {
+                file = await this.client.files.create({
+                    file: cf,
+                    purpose: 'assistants'
+                })
+            } catch (e) {
+                console.log(e)
+            }
+            addedFiles.push(file)
+        }
+
+        return addedFiles
+    }
```
Possible use of `Promise.all()`:

```
async addFiles(user: User, collection: string, fileData: any[]) {
    const filePromises = fileData.map(async (f) => {
        const splitDataURI = f.split(',');
        const fileName = getFileName(f);
        const updatedName = `${collection}_${String(user.id)}_${fileName}`;

        splitDataURI.pop();
        const bf = Buffer.from(splitDataURI.pop() || '', 'base64');
        const cf = new File([bf], updatedName);

        try {
            const file = await this.client.files.create({
                file: cf,
                purpose: 'assistants'
            });
            return file;
        } catch (e) {
            console.error(e);
            return null; // or handle the error as appropriate
        }
    });

    return Promise.all(filePromises);
}
```

innov8r-44 reviewed (approved):
Mostly LGTM, some places where we may be able to use `Promise.all` but this sort of depends on how many calls we expect in these cases. If there are eventually a lot it may be better to keep them synchronously .

mclean-connor commented:
made a few updates:
- added promise.all to file upload/delete methods. 
- added promise.all to a few endpoints in server index.json
- removed db entities for collections/collection files. it was only needed for openai since they only use files in one directory... not file collections. for now we can use file naming convention to specify file owner/collection of the file but will probably want to discuss updating this method.
